### PR TITLE
feat: enhance trades query to allow filtering for multiple markets an…

### LINKED ISSUE #7916: Enhance tradesConnection and trades subscription with TradeFilter
# Feature Overview

**In order to** Show all trades for multiple pubkeys/markets
**We will** lift the tradesConnection query to the top level
**And** add required filters to the tradesConnection query
**And** add required filters to the trades subscription
**So that** So that users can more easily see all their trades across multiple keys and markets

# Acceptance Criteria

- It is possible to filter trades by one or more marketIds
- It is possible to filter trades by one ore more partyIds
- any system tests that are broken due to this change are fixed (system test issue to be opened)

Note:
- Similar filters were added to the `positions` query
- The `positionsConnection` query was deprecated in favor of a top level `positions` query, perhaps we'd want to do the same here

# Example

```
// SCHEMA CHANGES
TradesFilter {
  partyIds: [ID!],
  marketIds: [ID!],
}

type Query {
  tradesConnection(
    filter: TradesFilter
    pagination: Pagination
    dateRange: DateRange
  ): TradesConnection
}

type Subscription {
  trades(
    filter: TradesFilter
  ): [TradeUpdate!]!
}
```

### PR DESCRIPTION
**Pull request by guoguojin**
…d parties etc

Closes #7916

> Note: This has a breaking change in the gRPC ListTradeRequest message which will also affect the REST API in order to support the passing of multiple marketIDs, partyIDs and orderIDs when filtering the trades.

### CODE REVIEW

File: datanode/gateway/graphql/schema.graphql [hunk 1]
```diff
@@ -656,18 +671,17 @@ type Query {
...
-  "List reward summary per epoch by asset, market, reward type"
-  epochRewardSummaries(
-    filter: RewardSummaryFilter,
-    pagination: Pagination
-  ): EpochRewardSummaryConnection
-
   "Get statistics about the Vega node"
   statistics: Statistics!
 
+  trades(
```
Comment (mattrussell36) [inline datanode/gateway/graphql/schema.graphql:677]: Missing a comment here for docs generation

**Review by mattrussell36 — DISMISSED** [review]

File: protos/sources/data-node/api/v2/trading_data.proto [hunk 2]
```diff
@@ -1590,11 +1590,11 @@ message GetLastTradeResponse {
 // Request to list trades
 message ListTradesRequest {
     // Market ID for which to list trades for, if provided
```
Comment (wwestgarth) [inline protos/sources/data-node/api/v2/trading_data.proto:1592]: Can you update all these comments now that the type has changed?

**Comment by wwestgarth** [comment]
Do we need to do the same for `ObserveTradesRequest{}`?

File: datanode/gateway/graphql/schema.graphql [hunk 3]
```diff
@@ -1784,7 +1794,7 @@ type Market {
     dateRange: DateRange,
     "Pagination information"
     pagination: Pagination
-  ): TradeConnection
+  ): TradeConnection @deprecated(reason: "Use root trades query instead")
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:1797]: Could you make the reason a bit clearer/easier to understand? Don't think the word "root" is adding a ton.

Comment (guoguojin) [reply datanode/gateway/graphql/schema.graphql]: Resolved https://github.com/vegaprotocol/vega/pull/7979/commits/17eaabb6243fb4c169fd2daaf687603388fb7e12

**Review by candida-d — DISMISSED** [review]

**Review by mattrussell36 — APPROVED** [review]

File: datanode/gateway/graphql/schema.graphql [hunk 4]
```diff
@@ -131,10 +131,8 @@ type Subscription {
 
   "Subscribe to the trades updates"
   trades(
-    "ID of the market from which you want trades updates"
-    marketId: ID
-    "ID of the party from which you want trades updates"
-    partyId: ID
+    "Filter to apply to trades"
+    filter: TradesSubscriptionFilter!
```
Comment (mattrussell36) [inline datanode/gateway/graphql/schema.graphql:135]: Can make this not a breaking change? Deprecate the current trades sub and add a new `tradesStream`?

**Review by mattrussell36 — CHANGES_REQUESTED** [review]

**Review by mattrussell36 — COMMENTED** [review]
Just making my review "comment only", so as not to block :)

File: datanode/gateway/graphql/schema.graphql (hunk 4 above)
Comment (guoguojin) [reply datanode/gateway/graphql/schema.graphql:135]: Sure, we can do that, it shouldn't be a problem. I'll make the changes.

Comment (guoguojin) [reply datanode/gateway/graphql/schema.graphql:143]: resolved with https://github.com/vegaprotocol/vega/pull/7979/commits/31ba0e6701ea3bd654043a339bf372008810be3f

**Review by mattrussell36 — DISMISSED** [review]

**Review by candida-d — DISMISSED** [review]

File: datanode/gateway/graphql/resolvers.go [hunk 5]
```diff
@@ -2193,9 +2216,65 @@ func (r *mySubscriptionResolver) Orders(ctx context.Context, filter *OrderByMark
...
+
+	if party != nil {
+		parties = append(parties, *party)
+	}
+
 	req := &v2.ObserveTradesRequest{
-		MarketId: market,
-		PartyId:  party,
+		MarketId: markets,
+		PartyId:  parties,
```
Comment (jeremyletang) [inline datanode/gateway/graphql/resolvers.go:2231]: Shouldn't these fields be names MarketIds and PartyIds?

File: datanode/gateway/graphql/resolvers.go [hunk 6]
```diff
@@ -2193,9 +2216,65 @@ func (r *mySubscriptionResolver) Orders(ctx context.Context, filter *OrderByMark
...
+	}()
+
+	return c, nil
+}
+
+func (r *mySubscriptionResolver) TradesStream(ctx context.Context, filter TradesSubscriptionFilter) (<-chan []*types.Trade, error) {
+	req := &v2.ObserveTradesRequest{
+		MarketId: filter.MarketIds,
+		PartyId:  filter.PartyIds,
 	}
```
[the remark this answers is not in the document]
Comment (guoguojin) [reply datanode/gateway/graphql/resolvers.go]: Resolved in https://github.com/vegaprotocol/vega/pull/7979/commits/df717796c53dc24d1408a3826a473e8d96f1e07d

File: datanode/gateway/graphql/resolvers.go (hunk 5 above)
Comment (guoguojin) [reply datanode/gateway/graphql/resolvers.go]: Resolved in https://github.com/vegaprotocol/vega/pull/7979/commits/df717796c53dc24d1408a3826a473e8d96f1e07d

**Review by candida-d — APPROVED** [review]

**Review by mattrussell36 — APPROVED** [review]

**Review by candida-d — APPROVED** [review]

**Review by ettec — APPROVED** [review]

**Pull request merged**